src/detect-metadata.c

```diff
@@ -28,23 +28,287 @@
+        HashTableFree(de_ctx->metadata_table);
+}
+
+static const char *DetectMedatataHashAdd(DetectEngineCtx *de_ctx, const char *string)
+{
+    const char * hstring = (char *)HashTableLookup(de_ctx->metadata_table, (void *)string, strlen(string));
+    if (hstring) {
+        return hstring;
+    }
+
+    const char *astring = SCStrdup(string);
+    if (astring == NULL) {
+        return NULL;
+    }
+
+    if (HashTableAdd(de_ctx->metadata_table, (void *)astring, strlen(astring)) == 0) {
+        return (char *)HashTableLookup(de_ctx->metadata_table, (void *)astring, strlen(astring));
+    } else {
+        SCFree((void *)astring);
+    }
+    return NULL;
+}
+
+static int DetectMetadataParse(DetectEngineCtx *de_ctx, Signature *s, const char *metadatastr)
+{
+#define MAX_SUBSTRINGS 30
+    int ret = 0, res = 0;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = pcre_exec(parse_regex, parse_regex_study, metadatastr,
+                    strlen(metadatastr), 0, 0, ov, MAX_SUBSTRINGS);
+    if (ret < 2) {
+        /* Only warn user that the metadata are not parsed due
+         * to invalid format */
+        SCLogInfo("signature metadata not in key value format, ret %" PRId32
+                    ", string %s", ret, metadatastr);
+        return 0;
+    }
+
+    char *saveptr = NULL;
+    size_t metadatalen = strlen(metadatastr)+1;
+    char rawstr[metadatalen];
+    strlcpy(rawstr, metadatastr, metadatalen);
+    char * kv = strtok_r(rawstr, ",", &saveptr);
+    const char *key = NULL;
+    const char *value = NULL;
+    char pkey[256];
+    char pvalue[256];
+
+    /* now check key value */
+    do {
+        DetectMetadata *dkv;
+
+        ret = pcre_exec(parse_tag_regex, parse_tag_regex_study, kv, strlen(kv), 0, 0, ov, MAX_SUBSTRINGS);
+        if (ret < 2) {
+            SCLogError(SC_ERR_PCRE_MATCH, "pcre_exec parse error, ret %" PRId32
+                    ", string %s", ret, rawstr);
+            goto error;
+        }
+
+        res =  pcre_copy_substring(kv, ov, MAX_SUBSTRINGS, 1, pkey, sizeof(pkey));
+        if (res < 0) {
+            SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_copy_substring failed");
+            goto error;
+        }
+        key = DetectMedatataHashAdd(de_ctx, pkey);
+        if (key == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "can't create metadata key");
+            goto error;
+        }
+
+        res =  pcre_copy_substring(kv, ov, MAX_SUBSTRINGS, 2, pvalue, sizeof(pvalue));
+        if (res < 0) {
+            SCLogError(SC_ERR_PCRE_GET_SUBSTRING, "pcre_copy_substring failed");
+            goto error;
+        }
+        value = DetectMedatataHashAdd(de_ctx, pvalue);
+        if (value == NULL) {
+            SCLogError(SC_ERR_MEM_ALLOC, "can't create metadata value");
+            goto error;
+        }
+
+        SCLogDebug("key: %s, value: %s", key, value);
+
+        dkv = SCMalloc(sizeof(DetectMetadata));
+        if (dkv) {
+            dkv->key = key;
+            if (!dkv->key) {
```

Comment:
this seems to be impossible: key != NULL because it is already checked at line 185